### PR TITLE
Chebychev shared roots

### PR DESCRIPTION
**Pull request by MSNetrom**
I have tried to implement a function-class that samples chebychev polynomials of same degree and same roots. The roots can be perturbed randomly and the sign of the function is set randomly. This is similar to what we have done in our experiments. I have also tried implementing everything in pytorch, in contrast to our original implementation.

I have not tried to train anything, except for testing that training runs.

### CODE REVIEW

File: src/function_classes/chebychev_specials.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+        super(ChebychevSharedRoots, self).__init__(*args, **kwargs)
+
+    def _init_param_dist(self) -> D.Distribution:
+        """Produce the distribution with which to sample parameters"""
+        perturbationd_dist = D.Uniform(-self.perturbation*torch.ones_like(self.chebychev_roots), self.perturbation*torch.ones_like(self.chebychev_roots))
+        return perturbationd_dist
+    
+    def evaluate(self, x_batch: torch.Tensor, params: torch.Tensor) -> torch.Tensor:
+
```
Comment (riensou) [inline src/function_classes/chebychev_specials.py:34]: match the other functions, pass in `*params`, then in the beginning of the function, add `weights, *_ = params`

**Review by riensou — COMMENTED** [review]
overall looks good so far, mostly just styling issues

File: src/function_classes/__init__.py [hunk 2]
```diff
@@ -11,13 +11,18 @@
...
+
 FUNCTION_CLASSES = {
     "linear regression" : LinearRegression,
     "linear classification" : LinearClassification,
     "sparse linear" : SparseLinearRegression,
     "quadradtic regression" : QuadraticRegression,
     "2 layer mlp regression" : MLPRegression,
-    "decision tree" : DecisionTreeRegression
+    "decision tree" : DecisionTreeRegression,
+    "chebychev shared roots" : ChebychevSharedRoots
```
[the remark this answers is not in the document]
Comment (MSNetrom) [reply src/function_classes/__init__.py]: Fixed this in new coomit

File: src/function_classes/chebychev_specials.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+class ChebychevSharedRoots(FunctionClass):
+
+    """
+    This class generates polynomials with shared roots
+    Roots can be perturbed
+    And the polynomials are mainly scaled so that the maximum value is 1
+    Although scaling_perc is the percentage of scaling that is uniform random
+    """
+
+    def __init__(self, degree, perturbation=0.1, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (MSNetrom) [reply src/function_classes/chebychev_specials.py:13]: Fixed this in new coomit

File: src/function_classes/chebychev_specials.py (hunk 1 above)
Comment (MSNetrom) [reply src/function_classes/chebychev_specials.py]: Fixed this in new coomit

**Comment by MSNetrom** [comment]
The "chebychev shared roots" and "chebychev kernel linear regression" has the same code now.
The test-function-classes is not successfull as this function-class were made for 1-dimensional input. That is (batch_size, seq_length, 1) instead of (batch_size, seq_length, any)

**Comment by riensou** [comment]
> The "chebychev shared roots" and "chebychev kernel linear regression" has the same code now. The test-function-classes is not successfull as this function-class were made for 1-dimensional input. That is (batch_size, seq_length, 1) instead of (batch_size, seq_length, any)

@nelson-lojo can we change checks.sh so that x_dim is not always 10, for instance in this case it needs to be 1? otherwise, this looks ready to merge to me

**Comment by nelson-lojo** [comment]
> @nelson-lojo can we change checks.sh so that x_dim is not always 10, for instance in this case it needs to be 1? otherwise, this looks ready to merge to me

Yeah I think a clean-to-use way is to have a thin parser that checks if there's a comment on the line with the function class. e.g.:
```python
FUNCTION_CLASSES = {
    ...
    "my_function_class" : MyFunctionClass, # x_dim 1
    ...
}
```
Where each value will have a default in `checks.sh`

🤔 Parsing in bash will be a fun challenge

**Review by nelson-lojo — APPROVED** [review]
lgtm! Minor spelling/style choices, but nothing that would change code that calls this.

**Comment by MSNetrom** [comment]
> lgtm! Minor spelling/style choices, but nothing that would change code that calls this.

Am I supposed to click merge, ir how does this work?

**Comment by nelson-lojo** [comment]
@MSNetrom It looks like you need to squash merge our main into your branch to pass CI -- the error showing up there was fixed only recently. As soon as the CI passes, I'll merge this in

**Pull request merged**